fix(core/webidl-index): support multiple idl blocks

shubhshrma opened the pull request:
For issue #1479 
The problem is appending only the first child.


Screenshot after changes: 

![peek 2018-02-11 20-30](https://user-images.githubusercontent.com/20224346/36074759-a3b4bd04-0f6a-11e8-9343-42efdf805faf.gif)


UPDATED SCREENSHOT

![peek 2018-02-11 23-24](https://user-images.githubusercontent.com/20224346/36076661-39e8d5ca-0f85-11e8-82c2-8bd897eff2b9.gif)

marcoscaceres commented:
Great find! Could you please check why the idl-index test is now failing tho? We might’ve also want to add the test case from the original bug to the test suite.

marcoscaceres commented on src/core/webidl-index.js:
```diff
@@ -42,8 +42,11 @@ export function run(conf, doc, cb) {
       const span = document.createElement("span");
-      const clone = elem.cloneNode(true).firstElementChild;
-      span.appendChild(clone);
+      const clone = elem.cloneNode(true).children;
```
Nit: `clone` is no longer s good name. Let’s go with:

```js
const [children] = element.cloneNode(true);
```

shubhshrma replied:
I've updated the code. Please see whether it's good now.

marcoscaceres commented on src/core/webidl-index.js:
```diff
@@ -42,8 +42,11 @@ export function run(conf, doc, cb) {
+      const clone = elem.cloneNode(true).children;
+      Array.from(clone).forEach(child => {
+        span.appendChild(child);
+        span.appendChild(document.createTextNode("\n"));
```
If a child is already a new line, should we won’t this give us multiple new lines?

shubhshrma replied:
Yes. This is giving an extra newline after every ```pre.idl```. Looking into this.

shubhshrma commented:
Regarding the tests, Travis passed now. I'll update the tests in some time @marcoscaceres.

marcoscaceres commented:
Awesome! Thanks @shubhshrma.

marcoscaceres commented:
Heh, autocorrect got me. I meant “quick”, not ”kick”, comments! :)

saschanaz commented on src/core/webidl-index.js:
```diff
@@ -42,9 +42,11 @@ export function run(conf, doc, cb) {
-      span.appendChild(clone);
-      span.appendChild(document.createTextNode("\n"));
+      const children = elem.cloneNode(true).children;
+      Array.from(children).forEach(child => {
```
Can we use the new for-of syntax? `for (const child of Array.from(children))`

marcoscaceres replied:
We’ve been using mostly functional style, but here it might make sense as it’s a single `forEach`.

marcoscaceres commented:
@shubhshrma, when possible, please rebase.

shubhshrma commented:
Sure I'll rebase @marcoscaceres. Um, I am facing a small problem writing tests in ```tests/spec/core/idl-index-spec.js```. I've written this:
```
it("allows multi-block idl", async () => {
    const body = `
      ${makeDefaultBody()}
      <section>
        <pre class=idl>
          [Constructor, Exposed=Window]
          interface BeforeInstallPromptEvent : Event {
              Promise&lt;PromptResponseObject&gt; prompt();
          };
          dictionary PromptResponseObject {
            AppBannerPromptOutcome userChoice;
          };
        </pre>
      </section>
      <section id="idl-index"></section>
    `;
    const expectedIDL = `[Constructor, Exposed=Window]
      interface BeforeInstallPromptEvent : Event {
          Promise&lt;PromptResponseObject&gt; prompt();
      };
      dictionary PromptResponseObject {
        AppBannerPromptOutcome userChoice;
      };\n`;
    const ops = {
      config: makeBasicConfig(),
      body,
    };
    const doc = await makeRSDoc(ops);
    var idlIndex = doc.querySelector("#idl-index");
    expect(idlIndex).not.toBe(null);
    expect(idlIndex.querySelector("pre").textContent).toEqual(expectedIDL);
  });
```
This is not passing because the entities ```&lt; &gt;``` are not escaped in ```expectedIDL```. Is there an escaping utility present?

marcoscaceres commented:
> This is not passing because the entities &lt; &gt; are not escaped in expectedIDL. Is there an escaping utility present?

Just change them to "<" and ">", and that should work, no?

marcoscaceres commented:
Sorry,  I mean change them in "expectedIDL".

shubhshrma commented:
Please review @marcoscaceres @saschanaz. Sorry for the delay. I didn't realize that I had to make ```expectedIDL``` exactly formatted like ```body``` in the test. Now I have rebased and added the test.

marcoscaceres commented on tests/spec/core/idl-index-spec.js:
```diff
@@ -39,6 +39,42 @@ interface Bar {
+    var idlIndex = doc.querySelector("#idl-index");
+    expect(idlIndex).not.toBe(null);
+    console.log(idlIndex.querySelector("pre").textContent);
+    console.log(expectedIDL);
```
Please remove console statements

marcoscaceres replied:
Better, convert these to expect statements.

marcoscaceres commented on tests/spec/core/idl-index-spec.js:
```diff
@@ -39,6 +39,42 @@ interface Bar {
+        };
+        </pre>
+      </section>
+         <section id="idl-index"></section>
```
Nit: make sure tgis section aligns with the one above.

shubhshrma commented:
Rebased and updated @marcoscaceres.

marcoscaceres commented:
Just waiting on Travis to give ok.

marcoscaceres commented:
@shubhshrma, please take a look at: https://github.com/w3c/respec/issues/1502

shubhshrma commented:
Looking into it @marcoscaceres

marcoscaceres commented:
going live now as v19.2.2
